hathor/transaction/block.py

```diff
@@ -141,6 +141,8 @@ def get_next_block_best_chain_hash(self) -> Optional[bytes]:
         """
         assert self.storage is not None
         meta = self.get_metadata()
+        if meta.voided_by:
+            return None
```

Comment:
I'm not sure what's the best approach here. Should we raise an exception? I feel this method should never be called in voided blocks because there's no such thing as next block in the best chain.

Reply:
I might have added this for testing, there is an assert immediately after, which I remember might have failed. I'm removing this and running all tests to see if this caused any problem.

Reply:
I'm not sure why, but the simulation tests fail too often when I remove this. This branch is may be introducing a bug that is causing this. I have to review it in depth.

Reply:
I think I got it. I'm adding this fix:

```diff
diff --git a/hathor/p2p/node_sync_v2.py b/hathor/p2p/node_sync_v2.py
index d26cd073..c8a5fc54 100644
--- a/hathor/p2p/node_sync_v2.py
+++ b/hathor/p2p/node_sync_v2.py
@@ -56,6 +56,7 @@ class StreamEnd(IntFlag):
     END_HASH_REACHED = 0
     NO_MORE_BLOCKS = 1
     LIMIT_EXCEEDED = 2
+    STREAM_BECAME_VOIDED = 3  # this will happen when the current chain becomes voided while it is being sent
 
 
 class NodeBlockSync(SyncManager):
@@ -924,6 +925,12 @@ class BlockchainStreaming(_StreamingBase):
             self.stop()
             self.node_sync.send_blocks_end(StreamEnd.LIMIT_EXCEEDED)
             return
+
+        if cur.get_metadata().voided_by:
+            self.stop()
+            self.node_sync.send_blocks_end(StreamEnd.STREAM_BECAME_VOIDED)
+            return
+
         self.counter += 1
 
         self.current_block = cur.get_next_block_best_chain()

```

What could happen is that while a peer is sending a stream of blocks to second peer, it is also receiving blocks from a third peer, such that a re-org happens and the chain that it was sending becomes voided. So the block we assumed to be part of the best chain, is now voided and would fail that assert. I'm restoring the assert and adding the check above on the sync-v2 class.

The simulator test that will give the most consistent failures on this case is `tests/simulation/test_simulator.py::SyncV2HathorSimulatorTestCase::test_new_syncing_peer`. I'm running it with:

```
pytest tests/simulation/test_simulator.py -k 'test_new_syncing_peer and v2' -s
```

With the fixes I mentioned it fails much less often, which I'm still debugging. Either way, I'm considering using flaky on some of these tests to actually require it to pass multiple times (we used flaky before to require a test to pass 1 out of 2 tries, but I can require it to pass 3 out of 3, for example).